### PR TITLE
Caffe2 Profiles

### PR DESCRIPTION
**Pull request by Maratyszcza**
Enable dynamic modification of OperatorDef attributes based on platform or SoC name

### CODE REVIEW

**Comment by Maratyszcza** [comment]
@caffe2bot retest this please

File: caffe2/core/profiles.cc [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    const cpuinfo_package* package = cpuinfo_get_package(0);
+    return std::string(package->name,
+      strnlen(package->name, CPUINFO_PACKAGE_NAME_MAX));
+  }
+};
+
+/**
+ * Checks if @a string starts with @a prefix.
+ */
+inline bool startsWith(
```
Comment (yinghai) [inline caffe2/core/profiles.cc]: Move to `utils/string_utils.h`?

Comment (Maratyszcza) [reply caffe2/core/profiles.cc]: Done!

**Comment by Maratyszcza** [comment]
@caffe2bot retest this please

**Review by Yangqing — CHANGES_REQUESTED** [review]
I think in offline discussions folks are having quite nontrivial concerns about this design, so we probably want to iterate on this before landing. cc @salexspb @ajtulloch @bwasti .

File: caffe2/proto/caffe2.proto [hunk 2]
```diff
@@ -219,6 +270,11 @@ message NetDef {
   // blobs' contents may be overwritten.
   repeated string external_input = 7;
   repeated string external_output = 8;
+
+  // List of profiles that modify operator arguments based on characteristics
+  // of the target system. Profiles are applied in the order in which they
+  // are listed.
+  repeated Profile profile = 9;
```
Comment (salexspb) [inline caffe2/proto/caffe2.proto:277]: I personally feel this is a hack. Protobufs are meant to be very rarely updated. The Predicate concept is very vague. We will have to keep adding stuff there if we decide to go with protobuf changing approach. So I would discourage emulating predicates on the protobuf level. 

I wonder what are technical difficulties you see when you implement this without any changes to the protobuf. So far engines + runtime dispatch based on the available hardware worked. You just need to write dispatch logic on the library code level, but not on the serialization level. For example, check out perfkernels code, this is dispatch Dima and Yangqing added last year to support fast embedding lookup kernels from Misha. 

https://github.com/caffe2/caffe2/tree/01827c153db96349745a544148c1ff0386c5ef9e/caffe2/perfkernels

You could also think about this as a technical challenge - write the code in the way that dispatch works at runtime based on the hardware you have. But without such heavy changes to the protobuf.

**Pull request closed without merging**